File: Programmers/Level2/괄호회전하기/괄호회전하기.py
from collections import deque

def solution(s):
    answer = 0
    queue = deque(s)

    for _ in range(len(s)):
        stack = []
        for i in queue:
            if not stack:
                stack.append(i)
            elif stack[-1] == '[' and i == ']':
                stack.pop()
            elif stack[-1] == '(' and i == ')':
                stack.pop()
            elif stack[-1] == '{' and i == '}':
                stack.pop()
            else:
                stack.append(i)

        queue.rotate(-1)    
        
        if not stack: answer += 1
            
    return answer

# deque 라이브러리 미사용
def solution(s):
    answer = 0
    s = list(s)
    
    for _ in range(len(s)):
        stack = []
        for i in s:
            if not stack:
                stack.append(i)
            elif stack[-1] == '[' and i == ']':
                stack.pop()
            elif stack[-1] == '(' and i == ')':
                stack.pop()
            elif stack[-1] == '{' and i == '}':
                stack.pop()
            else:
                stack.append(i)

        s.append(s.pop(0))
        
        if not stack: answer += 1
        
    return answer

File: Programmers/Level2/괄호회전하기/괄호회전하기.js
function solution(s) {
    let answer = 0;
    let queue = s.split("");

    for (let _ = 0; _ < s.length; _++) {
        let stack = [];
        for (let i = 0; i < queue.length; i++) {
            if (stack.length == 0) {
                stack.push(queue[i]);
                continue;
            }

            if (stack[stack.length - 1] === "[" && queue[i] === "]")
                stack.pop();
            else if (stack[stack.length - 1] === "(" && queue[i] === ")")
                stack.pop();
            else if (stack[stack.length - 1] === "{" && queue[i] === "}")
                stack.pop();
            else stack.push(queue[i]);
        }
        queue.push(queue.shift());

        if (stack.length === 0) answer += 1;
    }
    return answer;
}

// 다른 사람 풀이
function solution(s) {
    let answer = 0;
    for (let i = 0; i < s.length; i++) {
        const stack = [];
        const check = i === 0 ? s : s.slice(i) + s.slice(0, i);
        for (let j = 0; j < check.length; j++) {
            if (stack[stack.length - 1] === "(" && check[j] === ")")
                stack.pop();
            else if (stack[stack.length - 1] === "[" && check[j] === "]")
                stack.pop();
            else if (stack[stack.length - 1] === "{" && check[j] === "}")
                stack.pop();
            else stack.push(check[j]);
        }
        if (stack.length === 0) answer += 1;
    }
    return answer;
}
